Compute liveness constraints in location-sensitive polonius

lqd opened the pull request:
This continues the location-sensitive prototype. In this episode, we build the liveness constraints.

Reminder of the approach we're taking: we need variance data to create liveness edges in the forward/backward/both directions (respectively in the cases of covariance, contravariance, invariance) in the localized constraint graph.

This PR: 
- introduces the holder for that, and for the liveness data in the correct shape: the transpose of what we're using today, "live regions per points".
- records use/drop live region variance during tracing
- records regular live region variance at the end of liveness
- records the correctly shaped live region per point matrix
- uses all of the above to compute the liveness constraints

(There's still technically one tiny part of the liveness owl left to do, but I'll leave it for a future PR: we also need to disable the NLL optimization that avoids computing liveness for locals whose types contain a region outliving a free region -- the existing constraints make it effectively live at all points; this doesn't work under polonius)

r? @jackh726 cc @matthewjasper

lqd committed:
introduce polonius context

This context struct will hold data to help creating localized
constraints:
- the live regions, with the shape matching a CFG walk, indexed per
  point
- the variance of these live regions, represented as the direction we'll
  add the appropriate

We also add this structure to the mir typeck to record liveness data,
and make it responsible for localized constraint creation.

lqd committed:
add variance recording

Following the Generalizer's structure, relating e.g. a type with itself
will allow tracking the variance wrt the contained regions.

lqd committed:
record variance of use/drop live regions

records the variance of:
- use live types
- drop live generic args

lqd committed:
record variance of regular live regions

lqd committed:
improve `bit_set` assertion

it missed the index and bounds info

lqd committed:
finish filling polonius context

transpose liveness matrix and record live regions at the end of MIR typeck

lqd committed:
liveness constraints: draw the rest of the owl

lqd commented:
Again this should be gated and deactivated, but let's check if the changes have any impact: @bors try @rust-timer queue.

rust-timer commented:
Error occurred while parsing comment: Invalid command argument `.` (there may be no spaces around the `=` character)

bors commented:
:hourglass: Trying commit 47d46ae527fc5aef3b6857554d4a598e39a73f86 with merge df62f54b81cb778616d2f16730ff9e381f66fdec...
<!-- homu: {"type":"TryBuildStarted","head_sha":"47d46ae527fc5aef3b6857554d4a598e39a73f86","merge_sha":"df62f54b81cb778616d2f16730ff9e381f66fdec"} -->

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

compiler-errors commented on compiler/rustc_borrowck/src/polonius/liveness_constraints.rs:
```diff
@@ -0,0 +1,316 @@
+        T: Relate<TyCtxt<'tcx>>,
+    {
+        let result = self.relate(a.skip_binder(), a.skip_binder())?;
+        Ok(a.rebind(result))
```
probably should just be `Ok(a)`

compiler-errors replied:
oh, `a` isn't copy huh.

compiler-errors replied:
wait yes it is

lqd replied:
Cool, thank you. I wasn't exactly sure how close these should be to the `Generalizer`...

compiler-errors replied:
difference between the generalizer and this is that this doesn't actually return a meaningful value

lqd replied:
I fixed this in the latest push.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12459148470/job/34777930585)
Build commit: df62f54b81cb778616d2f16730ff9e381f66fdec (`df62f54b81cb778616d2f16730ff9e381f66fdec`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12459148470/job/34777930585"},"merge_sha":"df62f54b81cb778616d2f16730ff9e381f66fdec"} -->

rust-timer commented:
Queued df62f54b81cb778616d2f16730ff9e381f66fdec with parent e108481f74ff123ad98a63bd107a18d13035b275, future [comparison URL](https://perf.rust-lang.org/compare.html?start=e108481f74ff123ad98a63bd107a18d13035b275&end=df62f54b81cb778616d2f16730ff9e381f66fdec).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.9 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (df62f54b81cb778616d2f16730ff9e381f66fdec): [comparison URL](https://perf.rust-lang.org/compare.html?start=e108481f74ff123ad98a63bd107a18d13035b275&end=df62f54b81cb778616d2f16730ff9e381f66fdec&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=e108481f74ff123ad98a63bd107a18d13035b275&end=df62f54b81cb778616d2f16730ff9e381f66fdec&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.3% | [0.2%, 0.4%] | 3     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=e108481f74ff123ad98a63bd107a18d13035b275&end=df62f54b81cb778616d2f16730ff9e381f66fdec&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=e108481f74ff123ad98a63bd107a18d13035b275&end=df62f54b81cb778616d2f16730ff9e381f66fdec&stat=cycles:u)
<details>
<summary>Results (secondary -0.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.0%  | [2.0%, 2.0%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.9% | [-3.9%, -3.9%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=e108481f74ff123ad98a63bd107a18d13035b275&end=df62f54b81cb778616d2f16730ff9e381f66fdec&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 763.813s -> 762.245s (-0.21%)
**Artifact size**: 330.68 MiB -> 330.70 MiB (0.01%)

compiler-errors commented on compiler/rustc_borrowck/src/polonius/liveness_constraints.rs:
```diff
@@ -0,0 +1,315 @@
+    where
+        T: Relate<TyCtxt<'tcx>>,
+    {
+        Ok(a)
```
wait lol i think you misinterpreted my suggestion. It should be:

```suggestion
        self.relate(a.skip_binder(), a.skip_binder());
        Ok(a)
```

compiler-errors replied:
All I was saying was that you don't need to rebind the result

lqd replied:
The unfortunate thing is that there were no changes to tests due to this mistake.

compiler-errors replied:
yeah oof

lqd replied:
Should be good now. Merry xmas michael!

jackh726 commented on compiler/rustc_borrowck/src/polonius/mod.rs:
```diff
@@ -34,45 +34,84 @@
+
+    /// The expected edge direction per live region: the kind of directed edge we'll create as
+    /// liveness constraints depends on the variance of types with respect to each contained region.
+    live_region_variances: BTreeMap<RegionVid, ConstraintDirection>,
```
How "dense" is this? Could it make sense as an IndexVec? (Actually not sure what the speed tradeoffs are here.)

lqd replied:
It's usually rather sparse, for example:
- from a quick look at our entire test suite, 70% of functions have no liveness or variance here, more than 75% of functions have variances here for half of the regions; of course it's unlikely to be representative of real world code. 
- on a somewhat more realistic codebase, rustc-perf's primary benchmarks, numbers look similar but with a smaller magnitude, à la most functions have variance for <half of their regions.

That being said, I haven't started looking at performance yet. I can add a fixme if you want, to check the best data structure to use in the future. The number of regions is small, the access patterns are mostly random with some locality for close regions (in id, and in time on the cfg). We'd need to profile anyways.

jackh726 replied:
That's okay, no need for a FIXME.

jackh726 commented on compiler/rustc_borrowck/src/polonius/mod.rs:
```diff
@@ -34,45 +34,84 @@
+    /// - encoding liveness constraints
+    pub(crate) fn create_localized_constraints<'tcx>(
+        &self,
+        regioncx: &mut RegionInferenceContext<'tcx>,
```
Why is this &mut?

lqd replied:
In the full prototype, we mutate the live loans in regioncx, as they're part of the location-insensitive data there (and are then used to compute the borrows in scope to check for errors on invalidations). It's not yet necessary in this PR, so I've removed the mut. It should be possible to move these live loans to the polonius context anyways, as the location-insensitive data itself should be deleted soon as well.

jackh726 reviewed (commented):
First commit

jackh726 commented on compiler/rustc_borrowck/src/type_check/mod.rs:
```diff
@@ -181,6 +180,12 @@ pub(crate) fn type_check<'a, 'tcx>(
+    if let Some(polonius_context) = typeck.polonius_context.as_mut() {
+        let num_regions = infcx.num_region_vars();
+        let points_per_live_region = typeck.constraints.liveness_constraints.points();
+        polonius_context.record_live_regions_per_point(num_regions, points_per_live_region);
```
I wonder instead of using an `Option` in PoloniusContext, if it's better to just use a different struct for MirTypeckResults...

lqd replied:
Yes, however I'd like to do something similar in a follow-up if you don't mind: what we eventually want to do I think is compute the localized constraints earlier (as they should only depend on mir typeck + liveness), and only store loan liveness in the context instead, with the optional debugging data that the polonius mir dump needs.

jackh726 commented on compiler/rustc_borrowck/src/polonius/liveness_constraints.rs:
```diff
@@ -231,11 +231,9 @@ struct VarianceExtractor<'a, 'tcx> {
-
-        if region.is_bound() || region.is_erased() {
-            // ignore these
+        let Some(region) = self.universal_regions.try_to_region_vid(region) else {
```
Hmm, so this seems interesting. Where would we be seeing a `Bound` or `Erased` region?  Seems curious to me that there's some subset of regions that we're essentially just ignoring.

jackh726 replied:
Could be worth adding a comment explaining the cases this might happen.

lqd replied:
This came about like so: 
- I wanted the variance for vars and free regions and could only get it by duplicating the Generalizer's trick with `TypeRelation`, which unfortunately also yields regions that can't be converted to vids. 
- so I just ignored these ICE causing kinds and saw no particular changes in the tests (though there are still 6 failures left to diagnose and they could hiding in there): as far as I know there were no live regions that lacked variance due to this. There's also similarly [unexpected kinds reaching borrowck](https://github.com/search?q=repo%3Arust-lang%2Frust+cannot+convert+to+a+region+vid&type=issues), due to issues upstream and/or the ongoing effort of continuing compilation of code as deep as possible, that could be involved. To see whether we should be trying to e.g. recurse after skipping or discharging the binder, I casually brought it up with Niko who said "seems fine to ignore for now" and we didn't think more of it since then. 

Since then I also discovered our test coverage is actually very weak in that area, so I'll ask you what is the correct thing to do here?

jackh726 replied:
Personally, I would prefer to just fail to compile if we ever hit this (as we do for the examples you linked).

Really, I think there's an underlying problem here that we should investigate.

lqd replied:
Sure, I'll put the ICEs back in then, and we can recheck and investigate later.

jackh726 replied:
Yeah, for sure we can always come back to it.

lqd replied:
> Sure, I'll put the ICEs back in then, and we can recheck and investigate later.

Actually maybe not, I forgot there's too many of them and I don't believe they really matter here: we're not ignoring regions in the graph because of the (transparent to reachability) fallback if no variance is found for a region.  As for `'erased` though, it was in [this test]() that involves GCE amongst others, so imho it's fine to ignore for now.

A fixme like you initially suggested looks best under these circumstances. I'll ping you on zulip to discuss.

jackh726 replied:
A FIXME is fine.

lqd replied:
Ok cool, I've pushed a fixme in the last commit, explaining why it's fine to not have variance data for these for now but we still need to look into it.

jackh726 reviewed (approved):
Two small thoughts, but r=me with or without addressing them.

lqd committed:
address review comments

- add a FIXME when looking for the region variance of unexpected regions
- drive-by: fix a doc comment link
- drive-by: simplify the variance match using exported variants instead

jackh726 commented:
@bors r+

bors commented:
:pushpin: Commit 089c525df2410d3d2a99767c0fc93689a4432fed has been approved by `jackh726`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jackh726 089c525df2410d3d2a99767c0fc93689a4432fed -->
<!-- homu: {"type":"Approved","sha":"089c525df2410d3d2a99767c0fc93689a4432fed","approver":"jackh726","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 089c525df2410d3d2a99767c0fc93689a4432fed with merge 8cdc67ed23e486a9f2b3e667a50d9b8782ba4056...
<!-- homu: {"type":"BuildStarted","head_sha":"089c525df2410d3d2a99767c0fc93689a4432fed","merge_sha":"8cdc67ed23e486a9f2b3e667a50d9b8782ba4056"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12540697957/job/34971297975)
Approved by: jackh726
Pushing 8cdc67ed23e486a9f2b3e667a50d9b8782ba4056 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jackh726","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12540697957/job/34971297975"},"merge_sha":"8cdc67ed23e486a9f2b3e667a50d9b8782ba4056"} -->

rust-timer commented:
Finished benchmarking commit (8cdc67ed23e486a9f2b3e667a50d9b8782ba4056): [comparison URL](https://perf.rust-lang.org/compare.html?start=6cd33d889d0cd554e195d0eb9a38af88152b786f&end=8cdc67ed23e486a9f2b3e667a50d9b8782ba4056&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=6cd33d889d0cd554e195d0eb9a38af88152b786f&end=8cdc67ed23e486a9f2b3e667a50d9b8782ba4056&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.1% | [0.1%, 0.1%] | 3     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6cd33d889d0cd554e195d0eb9a38af88152b786f&end=8cdc67ed23e486a9f2b3e667a50d9b8782ba4056&stat=max-rss)
<details>
<summary>Results (secondary -2.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.8% | [-4.8%, -1.1%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=6cd33d889d0cd554e195d0eb9a38af88152b786f&end=8cdc67ed23e486a9f2b3e667a50d9b8782ba4056&stat=cycles:u)
<details>
<summary>Results (primary -0.8%, secondary -2.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.8% | [-0.8%, -0.8%] | 1     |
| Improvements ✅ <br /> (secondary) | -2.0% | [-2.0%, -2.0%] | 1     |
| All ❌✅ (primary)                 | -0.8% | [-0.8%, -0.8%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=6cd33d889d0cd554e195d0eb9a38af88152b786f&end=8cdc67ed23e486a9f2b3e667a50d9b8782ba4056&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 761.218s -> 760.358s (-0.11%)
**Artifact size**: 325.50 MiB -> 325.48 MiB (-0.00%)
